typescript/web/src/components/datasets/upsert-dataset-modal.tsx

```diff
@@ -102,8 +121,50 @@ export const UpsertDatasetModal = ({
         name: datasetName,
         workspaceSlug,
       },
-      refetchQueries: ["getDatasets"],
-      awaitRefetchQueries: true,
+      update: (cache, { data }) => {
+        if (data != null) {
+          const { createDataset } = data;
+          const datasetCacheResult = cache.readQuery<Pick<Query, "datasets">>({
+            query: getDatasetsQuery,
+            variables: { where: { workspaceSlug } },
+          });
+          if (datasetCacheResult?.datasets == null) {
+            throw new Error(`Error retrieving datasets`);
+          }
+
+          const { datasets } = datasetCacheResult;
+          const updatedDatasets = datasets.concat(createDataset);
+          cache.writeQuery({
+            query: getDatasetsQuery,
+            variables: { where: { workspaceSlug } },
+            data: { datasets: updatedDatasets },
+          });
+        } else {
+          throw new Error("Received null data during dataset creation");
+        }
+      },
```

Comment:
In a normal time, I'd have asked to move this block out of the component.

But, since:
1. You're not the author of this big block.
2. We're focusing on deleting #741 

We'll stay with this until we have more time to split the UpsertDatasetModal component into smaller parts 🚀